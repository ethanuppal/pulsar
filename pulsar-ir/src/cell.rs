//! Copyright (C) 2024 Ethan Uppal. This program is free software: you can
//! redistribute it and/or modify it under the terms of the GNU General Public
//! License as published by the Free Software Foundation, either version 3 of
//! the License, or (at your option) any later version.

use std::fmt::{self, Display};

use crate::memory::Memory;
use pulsar_frontend::ast::ty::{LiquidTypeValue, Type, TypeValue};

/// A hardware element capable of storage.
pub enum Cell {
    Memory(Memory),

    /// `Cell::Register(bit_width)` is a register `bit_width` bits wide.
    Register(usize)
}

impl Cell {
    pub fn from(ty: &Type) -> Self {
        match ty.value {
            TypeValue::Unit => Self::Register(0),
            TypeValue::Var(_) => panic!(),
            TypeValue::Name(_) => todo!(),
            TypeValue::Int64 => Self::Register(64),
            TypeValue::Array(element_type, element_count) => {
                let LiquidTypeValue::Equal(length) = element_count.value else {
                    panic!("liquid type not resolved");
                };
                Self::Memory(Memory::new(length, element_type.size(), 1))
            }
            TypeValue::Function {
                inputs: _,
                outputs: _
            } => panic!()
        }
    }
}

impl Display for Cell {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Cell::Memory(memory) => {
                write!(
                    f,
                    "Memory(length={}, element={}, bank={})",
                    memory.length, memory.element, memory.bank
                )
            }
            Cell::Register(bit_width) => {
                write!(f, "Register(width={})", bit_width)
            }
        }
    }
}
