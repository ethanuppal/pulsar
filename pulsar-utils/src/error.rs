//! Copyright (C) 2024 Ethan Uppal. This program is free software: you can
//! redistribute it and/or modify it under the terms of the GNU General Public
//! License as published by the Free Software Foundation, either version 3 of
//! the License, or (at your option) any later version.

use super::loc::{Span, SpanProvider};
use colored::*;
use std::{
    fmt::{self, Display},
    io
};

#[repr(i32)]
#[derive(Clone, Copy, Debug)]
pub enum ErrorCode {
    WompWomp,
    UnrecognizedCharacter,
    UnexpectedEOF,
    UnexpectedToken,
    InvalidTopLevelConstruct,
    ConstructShouldBeTopLevel,
    InvalidTokenForStatement,
    InvalidOperatorSyntax,
    MalformedType,
    UnboundName,
    StaticAnalysisIssue
}

impl Display for ErrorCode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        (*self as i32).fmt(f)
    }
}

impl Default for ErrorCode {
    fn default() -> Self {
        Self::WompWomp
    }
}

#[derive(PartialEq, Eq, Debug)]
pub enum Level {
    Info,
    Note,
    Warning,
    Error
}

impl Level {
    fn color(&self) -> Color {
        match self {
            Self::Info => Color::BrightWhite,
            Self::Note => Color::BrightYellow,
            Self::Warning => Color::Yellow,
            Self::Error => Color::BrightRed
        }
    }

    fn form_header(&self, code: ErrorCode) -> ColoredString {
        let string = self.to_string();
        format!(
            "{}[{}{:04}]",
            string,
            string.chars().next().unwrap().to_ascii_uppercase(),
            code
        )
        .color(self.color())
    }
}

impl Display for Level {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Info => "info",
                Self::Note => "note",
                Self::Warning => "warning",
                Self::Error => "error"
            }
        )
    }
}

#[derive(PartialEq, Eq, Debug)]
pub enum Style {
    Primary,
    Secondary
}

/// An error at a given location with various information and diagnostics. Use
/// [`Error::fmt`] to obtain a printable version of the error.
#[derive(Debug)]
pub struct Error {
    style: Style,
    level: Level,
    code: ErrorCode,
    span: Option<Span>,
    message: String,
    explain: Option<String>,
    fix: Option<String>
}

impl Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // Only primary-style messages have a header indicating they are the
        // root of an error message and not auxillary information
        if self.style == Style::Primary {
            write!(f, "{}: ", self.level.form_header(self.code).bold(),)?;
            if self.span.is_some() {
                write!(
                    f,
                    "{}: ",
                    self.span.as_ref().unwrap().start.to_string().underline()
                )?;
            }
        }
        writeln!(f, "{}", self.message.bold())?;

        // If there is no span associated with this error, then we have
        // nothing more to print
        if self.span.is_none() {
            return Ok(());
        }

        // Otherwise, we print the span via a sequence of lines from the
        // source.
        let span = self.span.as_ref().unwrap();
        let span_extra_lines = (span.end.line - span.start.line) as usize;
        let show_lines_before = 1;
        let show_lines_after = 1;
        let mut already_explained = false;
        writeln!(f, "{}", "     │  ".dimmed())?;
        let (lines, current_line_pos) = span
            .start
            .lines(show_lines_before, span_extra_lines + show_lines_after);
        let show_start_line = span.start_line() - (show_lines_before as isize);
        let span_line_sections =
            span.find_intersection(&lines, show_start_line);
        for (i, line) in lines.iter().enumerate() {
            if i > 0 {
                writeln!(f)?;
            }
            write!(
                f,
                "{}",
                format!(
                    "{: >4} │  ",
                    i + (span.start.line as usize) - current_line_pos
                )
                .dimmed()
            )?;
            if i >= current_line_pos && i <= current_line_pos + span_extra_lines
            {
                let line_section = &span_line_sections[i - current_line_pos];
                let split_first = line_section.start;
                let (part1, rest) = line.split_at(split_first as usize);
                if !line.is_empty() {
                    let split_second =
                        split_first + (line_section.length() as isize) - 1;
                    let (part2, part3) = if (split_second as usize)
                        == line.len()
                    {
                        (rest, "")
                    } else {
                        rest.split_at((split_second - split_first + 1) as usize)
                    };
                    write!(
                        f,
                        "{}{}{}",
                        part1,
                        part2.color(self.level.color()),
                        if part3.is_empty() {
                            " ".on_color(self.level.color())
                        } else {
                            part3.into()
                        }
                    )?;
                } else {
                    write!(f, "{}{}", part1, " ".on_color(self.level.color()))?;
                }
                if let Some(explain) =
                    &self.explain.as_ref().filter(|_| !already_explained)
                {
                    already_explained = true;
                    fn create_error_pointer(length: usize) -> String {
                        match length {
                            0 => "".into(),
                            n => format!("└{}", "─".repeat(n - 1))
                        }
                    }
                    writeln!(f)?;
                    write!(
                        f,
                        "{}  {}{} {}",
                        "     │".dimmed(),
                        " ".repeat(part1.len()),
                        create_error_pointer(line_section.length())
                            .color(self.level.color()),
                        explain.bold().italic()
                    )?;
                }
            } else {
                write!(f, "{}", line)?;
            }
        }
        write!(f, "\n{}", "     │  ".dimmed())?;
        if let Some(fix) = &self.fix {
            write!(f, "\nSuggestion: {}", fix.bold())?;
        }
        Ok(())
    }
}

impl Default for Error {
    fn default() -> Self {
        Error {
            style: Style::Primary,
            level: Level::Error,
            code: ErrorCode::default(),
            span: None,
            message: String::default(),
            explain: None,
            fix: None
        }
    }
}

/// An interface for fluently constructing errors.
#[derive(Default)]
pub struct ErrorBuilder {
    error: Error
}

impl ErrorBuilder {
    /// Constructs a new error builder.
    pub fn new() -> Self {
        Self::default()
    }

    /// Uses the display style `style`.
    pub fn of_style(mut self, style: Style) -> Self {
        self.error.style = style;
        self
    }

    /// Uses the severity level `level`.
    pub fn at_level(mut self, level: Level) -> Self {
        self.error.level = level;
        self
    }

    /// Uses the error code `code`.
    pub fn with_code(mut self, code: ErrorCode) -> Self {
        self.error.code = code;
        self
    }

    /// Locates the error at the span given by `span_provider`.
    pub fn span<P: SpanProvider, PRef: AsRef<P>>(
        mut self, span_provider: PRef
    ) -> Self {
        self.error.span = Some(span_provider.as_ref().span());
        self
    }

    /// Identifies the error as having no location.
    pub fn without_loc(mut self) -> Self {
        self.error.span = None;
        self
    }

    /// Uses the main error description `message`.
    pub fn message(mut self, message: String) -> Self {
        self.error.message = message;
        self
    }

    /// Marks this error as without a message and instead continuing a previous
    /// error.
    ///
    /// Requires: the error is of secondary style, which must be set before this
    /// function is called (see [`ErrorBuilder::of_style`]).
    pub fn continues(self) -> Self {
        assert!(self.error.style == Style::Secondary);
        self.message("   ...".into())
    }

    /// Uses an explanatory note `explain`.
    pub fn explain(mut self, explain: String) -> Self {
        self.error.explain = Some(explain);
        self
    }

    /// Uses a recommendation `fix`.
    pub fn fix(mut self, fix: String) -> Self {
        self.error.fix = Some(fix);
        self
    }

    pub fn maybe_fix<S: AsRef<str>>(mut self, fix: Option<S>) -> Self {
        self.error.fix = fix.map(|str| str.as_ref().to_string());
        self
    }

    /// Produces the error.
    pub fn build(self) -> Error {
        self.error
    }
}

/// A shared error manager with an error recording limit.
pub struct ErrorManager {
    max_count: usize,
    primary_count: usize,
    errors: Vec<Error>
}

impl ErrorManager {
    /// Constructs an error manager that can record up to `max_count` primary
    /// errors.
    pub fn with_max_count(max_count: usize) -> ErrorManager {
        ErrorManager {
            max_count,
            primary_count: 0,
            errors: vec![]
        }
    }

    /// Whether the error manager has recorded an error.
    pub fn has_errors(&self) -> bool {
        !self.errors.is_empty()
    }

    /// Whether the error manager cannot take any further primary errors.
    pub fn is_full(&self) -> bool {
        self.primary_count == self.max_count
    }

    /// Records `error` and returns `true` unless `is_full()`.
    pub fn record(&mut self, error: Error) -> bool {
        if self.is_full() {
            false
        } else {
            if error.style == Style::Primary && error.level == Level::Error {
                self.primary_count += 1
            }
            self.errors.push(error);
            true
        }
    }

    /// Prints and clears all recorded errors to `output`.
    pub fn consume_and_write<W: io::Write>(
        &mut self, output: &mut W
    ) -> io::Result<()> {
        for (i, error) in self.errors.iter().enumerate() {
            if error.style == Style::Primary && i > 0 {
                output.write_all(&[b'\n'])?;
            }
            output.write_all(error.to_string().as_bytes())?;
            output.write_all(&[b'\n'])?;
            output.flush()?;
        }
        self.errors.clear();
        self.primary_count = 0;
        Ok(())
    }
}
