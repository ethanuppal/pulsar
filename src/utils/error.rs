use super::loc::Loc;
use colored::*;
use std::{fmt::Display, io};

type ErrorCode = i32;

pub enum Level {
    Info,
    Note,
    Warning,
    Error
}

impl Level {
    fn color(&self) -> Color {
        match self {
            Self::Info => Color::BrightWhite,
            Self::Note => Color::BrightYellow,
            Self::Warning => Color::Yellow,
            Self::Error => Color::BrightRed
        }
    }

    fn form_header(&self, code: ErrorCode) -> ColoredString {
        let string = self.to_string();
        format!(
            "{}[{}{:04}]",
            string,
            string.chars().next().unwrap().to_ascii_uppercase(),
            code
        )
        .color(self.color())
    }
}

impl Display for Level {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Info => "info",
                Self::Note => "note",
                Self::Warning => "warning",
                Self::Error => "error"
            }
        )
    }
}

pub enum Style {
    Primary,
    Secondary
}

/// An error at a given location with various information and diagnostics. Use
/// [`Error::fmt`] to obtain a printable version of the error.
pub struct Error {
    style: Style,
    level: Level,
    code: ErrorCode,
    loc: Loc,
    length: usize,
    message: String,
    explain: Option<String>,
    fix: Option<String>
}

impl Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}: {}\n",
            self.level.form_header(self.code).bold(),
            self.message.bold()
        )?;
        let (lines, current_line_pos) = self.loc.lines(2, 2);
        for (i, line) in lines.iter().enumerate() {
            if i > 0 {
                writeln!(f)?;
            }
            write!(
                f,
                "{}",
                format!("{: >4} │  ", i + self.loc.line - current_line_pos)
                    .dimmed()
            )?;
            if i == current_line_pos {
                let split_first = self.loc.col - 1;
                let split_second = split_first + self.length - 1;
                let (part1, rest) = line.split_at(split_first);
                let (part2, part3) =
                    rest.split_at(split_second - split_first + 1);
                write!(
                    f,
                    "{}{}{}",
                    part1,
                    part2.color(self.level.color()),
                    part3
                )?;
                if let Some(explain) = &self.explain {
                    fn create_error_pointer(length: usize) -> String {
                        match length {
                            0 => "".into(),
                            1 => "│".into(),
                            n => format!("└{}", "─".repeat(n - 1))
                        }
                    }
                    writeln!(f)?;
                    write!(
                        f,
                        "        {}{} {}",
                        " ".repeat(part1.len()),
                        create_error_pointer(part2.len())
                            .color(self.level.color()),
                        explain.bold().italic()
                    )?;
                }
            } else {
                write!(f, "{}", line)?;
            }
        }
        if let Some(fix) = &self.fix {
            write!(f, "\n{}", fix.bold())?;
        }
        Ok(())
    }
}

impl Default for Error {
    fn default() -> Self {
        Error {
            style: Style::Primary,
            level: Level::Error,
            code: ErrorCode::default(),
            loc: Loc::default(),
            length: 0,
            message: String::default(),
            explain: None,
            fix: None
        }
    }
}

pub struct ErrorBuilder {
    error: Error
}

impl ErrorBuilder {
    pub fn new() -> Self {
        ErrorBuilder {
            error: Error::default()
        }
    }

    pub fn of_style(mut self, style: Style) -> Self {
        self.error.style = style;
        self
    }

    pub fn at_level(mut self, level: Level) -> Self {
        self.error.level = level;
        self
    }

    pub fn at_region(mut self, loc: Loc, length: usize) -> Self {
        self.error.loc = loc;
        self.error.length = length;
        self
    }

    pub fn message(mut self, message: String) -> Self {
        self.error.message = message;
        self
    }

    pub fn explain(mut self, explain: String) -> Self {
        self.error.explain = Some(explain);
        self
    }

    pub fn fix(mut self, fix: String) -> Self {
        self.error.fix = Some(fix);
        self
    }

    pub fn build(self) -> Error {
        self.error
    }
}

pub struct ErrorManager {
    max_count: usize,
    errors: Vec<Error>
}

impl ErrorManager {
    fn with_max_count(max_count: usize) -> ErrorManager {
        ErrorManager {
            max_count,
            errors: vec![]
        }
    }

    fn push(&mut self, error: Error) -> bool {
        if self.errors.len() == self.max_count {
            false
        } else {
            self.errors.push(error);
            true
        }
    }

    fn consume<W: io::Write>(&mut self, mut output: W) -> io::Result<()> {
        for error in &self.errors {
            output.write_all(error.to_string().as_bytes())?;
            output.flush()?;
        }
        self.errors.clear();
        Ok(())
    }
}
