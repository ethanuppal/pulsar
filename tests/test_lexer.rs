#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use pulsar::utils::error::ErrorManager;
    use pulsar::{frontend::lexer::Lexer, utils::loc::Source};
    use std::cell::RefCell;
    use std::fs;
    use std::rc::Rc;

    fn read(filename: &str) -> Rc<Source> {
        Source::file(
            filename.into(),
            fs::read_to_string(filename)
                .expect(format!("Could not read file: {}", filename).as_str())
        )
    }

    fn lexer_output(
        filename: &str, error_manager: Rc<RefCell<ErrorManager>>
    ) -> String {
        let source = read(filename);
        let lexer = Lexer::new(source, error_manager);
        let mut output = String::new();
        for token in lexer {
            output.push_str(&format!("{:?}\n", token));
        }
        output
    }

    #[test]
    fn test_lexer() {
        let error_manager = ErrorManager::with_max_count(5);
        assert_snapshot!(lexer_output(
            "tests/data/lexer_in1.pl",
            error_manager.clone()
        ));
        assert_snapshot!(lexer_output(
            "tests/data/lexer_in2.pl",
            error_manager.clone()
        ));
        assert_snapshot!(lexer_output(
            "tests/data/lexer_in3.pl",
            error_manager.clone()
        ));
    }
}
