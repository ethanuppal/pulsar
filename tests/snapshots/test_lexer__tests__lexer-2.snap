---
source: tests/test_lexer.rs
expression: "lexer_output(\"tests/data/lexer_in2.pl\", error_manager.clone())"
---
(1, ty = integer, loc = tests/data/lexer_in2.pl:1:1)
(2, ty = integer, loc = tests/data/lexer_in2.pl:1:3)
(3, ty = integer, loc = tests/data/lexer_in2.pl:1:5)
(4, ty = integer, loc = tests/data/lexer_in2.pl:1:7)
(5, ty = integer, loc = tests/data/lexer_in2.pl:1:9)
(\n, ty = newline, loc = tests/data/lexer_in2.pl:1:10)
(id, ty = identifier, loc = tests/data/lexer_in2.pl:2:1)
(\n, ty = newline, loc = tests/data/lexer_in2.pl:2:3)
(id1, ty = identifier, loc = tests/data/lexer_in2.pl:3:1)
(\n, ty = newline, loc = tests/data/lexer_in2.pl:3:4)
(_id1, ty = identifier, loc = tests/data/lexer_in2.pl:4:1)
(\n, ty = newline, loc = tests/data/lexer_in2.pl:4:5)
(1, ty = integer, loc = tests/data/lexer_in2.pl:5:1)
(id, ty = identifier, loc = tests/data/lexer_in2.pl:5:2)
(\n, ty = newline, loc = tests/data/lexer_in2.pl:5:4)
